server/auth/create_token.go

```diff
@@ -0,0 +1,96 @@
+package auth
+
+import (
+	"encoding/json"
+	"github.com/gmo-personal/coding_challenge/server/csrf"
+	"github.com/gmo-personal/coding_challenge/server/utils"
+	"time"
+
+	"github.com/gmo-personal/coding_challenge/database"
+	"github.com/gmo-personal/coding_challenge/model"
+	"golang.org/x/crypto/bcrypt"
+	"net/http"
+)
+
+// Structure of a session, containing account ID and expiry time.
+type Session struct {
+	AccountID string
+	Expires   time.Time
+}
+
+// In memory sessions table to keep track of active sessions.
+// TODO This should be a database table to avoid memory and concurrency issues as well as scaling.
+var Sessions = make(map[string]Session)
+
+// Generates auth and CSRF token and returns them to the client.
+func PostTokenHandler(w http.ResponseWriter, r *http.Request) {
+	// Attempts to read the body, if unable to do so, return bad request.
+	accountJson, err := utils.GetBody(r)
+	if err != nil {
+			utils.LogError(err)
+			utils.ServeBadRequest(w)
+			return
+	}
+
+	// Attempts to parse json into account model, if unable to do so, return bad request.
+	account := &model.Account{}
+	err = json.Unmarshal(accountJson, account)
+	if err != nil {
+			utils.LogError(err)
+			utils.ServeBadRequest(w)
+			return
+	}
+
+	// Attempts to find an existing account by username and compares input and existing passwords,
+	// Serves unauthorized if they do not match.
+	existingAccount, err := database.SelectAccountByUsername(account.Username)
+	if err != nil {
+			utils.LogError(w)
+			utils.ServeNotFound(w)
+			return
+	}
+	if existingAccount != nil {
+		// Uses bcrypt library to compare the passwords, returns unauthorized if comparison fails.
+		err = bcrypt.CompareHashAndPassword([]byte(existingAccount.Password), []byte(account.Password))
+		if err != nil {
+				utils.LogError(w)
+				utils.ServeUnauthorized(w)
+				return
+		}
+		// Attempts to create an auth token for the account, serves internal server error on fail.
+		authToken, err := createAuth(existingAccount.ID)
+		if err != nil {
+				utils.LogError(w)
+				utils.ServeInternalServerError(w)
+				return
+		}
+		// Sends response to save the cookie.
+		utils.SaveCookie(authCookieName, authToken, w)
+
+		w.Header().Set("Content-Type", "application/json")
+
+		// Attempts to create an CSRF token for the account, serves internal server error on fail.
+		CSRFToken, err := csrf.CreateCSRF(w, r)
```

Comment:
You would prefer to have csrf token available even for public endpoints. So it's better to add this value to HTML template. It will be returned to the client on initial render where you would add it to the cookies as well.

```
<meta name="xcsrf_token" content="{{.Token}}">
```

Reply:
`<meta content="{{ .CSRFToken }}" name="csrf-token" />`
`RenderTemplate(w, "dashboard", &model.DashboardPage{CSRFToken: CSRFToken})`

The csrf token is injected to the html template here (dashboard handler)

Could you clarify? Thank you.

Reply:
I guess I was confused with below code where you retrieve CSRF token from the `/token` endpoint and then assign it to   `document.getElementById('X-CSRF-Token-input').value` . Why do you need to do that if you already injected this token in HTML? You do not need to regenerate it on login. 
 
```js
return await fetch('https://localhost:443/token/', {
        method: 'POST',
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        },
        body: userJson,
    }).then(async function(response) {
        if (response.ok) {
            console.log('submit passed')
            document.getElementById('X-CSRF-Token-input').value = await response.json();
            await login();
        } else {
            console.log('submit failed')
        }
        return response
    }).catch(function(e) {
        return null;
    })
```    

On a side node, I suggest to rewrite this function because it mixes up promises with async functions. Pick one way for dealing with asynchronous operations.

Reply:
I will look at the function. 

As for the above part. My intended login is two step: Login page on visiting site => user enters information and clicks login => handleSubmit() posts for tokens, CSRF token is put back into the login page on hidden element (and cookies) => login() occurs => server validates auth that we just received + token to ensure that we are logging in from the real login page => dashboard rendered with injected CSRF token. Is this unecessary?

I could skip this and just have an injected CSRF in the dashboard?

Reply:
> I could skip this and just have an injected CSRF in the dashboard?

Yes. Just make sure that every HTML page has CSRF token injected in it so JS code can pick it up to send back on `fetch` requests.

Reply:
Understood !